### PR TITLE
GODRIVER-2239 use source of PyKMIP

### PR DESCRIPTION
**Pull request by kevinAlbs**
GODRIVER-2239

# Summary
- Install source of PyKMIP instead of fork.
- Configure `kms_kmip_server.py` with new certificates to support default Golang TLS 1.2 cipher suites.
- Add scripts to `/etc` used to generate new Elliptic Curve key certificates.

Full patch build: https://evergreen.mongodb.com/version/6302ee619ccd4e09b5df281c

# Background & Motivation

Golang TLS 1.2 default ciphers are listed under [cipherSuitesPreferenceOrder here](https://go.dev/src/crypto/tls/cipher_suites.go).

The TLS cipher suites configured for PyKMIP TLS 1.2 are [listed here](https://github.com/OpenKMIP/PyKMIP/blob/0a3e39143f541775a4f8d0c6398069c4b412f24b/kmip/services/auth.py#L198).

There are two TLS cipher suites shared:
- `TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256`
- `TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384`

These cipher suites require Elliptic Curve, not RSA keys.

A PR was submitted to PyKMIP to add an RSA cipher suite supported by Golang: https://github.com/OpenKMIP/PyKMIP/pull/659. It has not been accepted.

Since this only impacts Golang, the certificates are added in the Go driver repo rather than driver-evergreen-tools.

### CODE REVIEW

File: .evergreen/config.yml [hunk 1]
```diff
@@ -481,8 +481,8 @@ functions:
           export AWS_DEFAULT_REGION="us-east-1"
 
           # Set client-side encryption credentials.
-          export CSFLE_TLS_CA_FILE="$DRIVERS_TOOLS/.evergreen/x509gen/ca.pem"
-          export CSFLE_TLS_CERTIFICATE_KEY_FILE="$DRIVERS_TOOLS/.evergreen/x509gen/client.pem"
+          export CSFLE_TLS_CA_FILE="$PROJECT_DIRECTORY/data/kmip-certs/ca-ec.pem"
```
Comment (prestonvasquez) [inline .evergreen/config.yml:484]: What does `$PROJECT_DIRECTORY` reference? Does this correspond with an update to the evergreen host builds?

File: etc/gen-ec-certs/gen-ec-certs.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# PyKMIP does not support Golang's default TLS cipher suites with RSA.
+# See: GODRIVER-G2239. 
+set -euo pipefail
+CA_SERIAL=$RANDOM
+SERVER_SERIAL=$RANDOM
+CLIENT_SERIAL=$RANDOM
+DAYS=14600
+
+# Generate CA certificate ... begin
+# Generate a EC private key.
```
Comment (prestonvasquez) [inline etc/gen-ec-certs/gen-ec-certs.sh:12]:
```suggestion
# Generate an EC private key.
```

File: etc/gen-ec-certs/gen-ec-certs.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+openssl ecparam -name prime256v1 -genkey -out ca-ec.key -noout
+# Generate a certificate signing request.
+openssl req -new -key ca-ec.key -out ca-ec.csr -subj "/C=US/ST=New York/L=New York City/O=MongoDB/OU=DBX/CN=ca/" -config empty.cnf -sha256
+# Self-sign the request.
+openssl x509 -in ca-ec.csr -out ca-ec.pem -req -signkey ca-ec.key -days $DAYS -sha256 -set_serial $CA_SERIAL
+# Generate CA certificate ... end
+
+# Generate Server certificate ... begin
+# Generate a EC private key.
```
Comment (prestonvasquez) [inline etc/gen-ec-certs/gen-ec-certs.sh:21]:
```suggestion
# Generate an EC private key.
```

File: etc/gen-ec-certs/gen-ec-certs.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# Generate a certificate signing request.
+openssl req -new -key server-ec.key -out server-ec.csr -subj "/C=US/ST=New York/L=New York City/O=MongoDB/OU=DBX/CN=server/" -config empty.cnf -sha256
+# Sign the request with the CA. Add server extensions.
+openssl x509 -in server-ec.csr -out server-ec.pem -req -CA ca-ec.pem -CAkey ca-ec.key -days $DAYS -sha256 -set_serial $SERVER_SERIAL -extfile server.ext
+# Append private key to .pem file.
+cat server-ec.key >> server-ec.pem
+# Generate Server certificate ... end
+
+# Generate Client certificate ... begin
+# Generate a EC private key.
```
Comment (prestonvasquez) [inline etc/gen-ec-certs/gen-ec-certs.sh:32]:
```suggestion
# Generate an EC private key.
```

File: etc/gen-ec-certs/gen-ec-certs.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+openssl req -new -key client-ec.key -out client-ec.csr -subj "/C=US/ST=New York/L=New York City/O=MongoDB/OU=DBX/CN=client/" -config empty.cnf -sha256
+# Sign the request with the CA. Add client extensions.
+openssl x509 -in client-ec.csr -out client-ec.pem -req -CA ca-ec.pem -CAkey ca-ec.key -days $DAYS -sha256 -set_serial $CLIENT_SERIAL -extfile client.ext
+# Append private key to .pem file.
+cat client-ec.key >> client-ec.pem
+# Generate Client certificate ... end
+
+# Clean-up.
+rm *.csr
+rm *.key
```
Comment (prestonvasquez) [inline etc/gen-ec-certs/gen-ec-certs.sh:45]: GitHub is implying there should be a new line at the end of the script. Does this sound right to you?

File: etc/gen-ec-certs/gen-ec-certs.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# This script is used to generate Elliptic Curve (EC) certificates.
+# The EC certificates are used for testing the Go driver with PyKMIP.
+# PyKMIP does not support Golang's default TLS cipher suites with RSA.
+# See: GODRIVER-G2239.
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply etc/gen-ec-certs/gen-ec-certs.sh]: Good eye. Thank you for the suggestions.

File: etc/gen-ec-certs/gen-ec-certs.sh (hunk 5 above)
Comment (kevinAlbs) [reply etc/gen-ec-certs/gen-ec-certs.sh]: Yes. Added a newline.

**Review by prestonvasquez — APPROVED** [review]

**Review by matthewdale — APPROVED** [review]
Looks good 👍

**Pull request merged**